ophyd/areadetector/cam.py

```diff
@@ -299,6 +303,80 @@ class EmergentVisionDetectorCam(CamBase):
     auto_gain = ADCpt(SignalWithRBV, 'EVTAutoGain')
 
 
+class EigerDetectorCam(CamBase, FileBase):
```

Comment:
Thanks!

Have you cross-referenced this with the implementation hard-coded into CHX's profile? I would bet that the implementation in this PR is more thoughtfully composed, but it would be good to know if it is a strict superset of the Components included there. Ideally we would move CHX to rely on this one instead.

https://github.com/NSLS-II-CHX/profile_collection/blob/7ef674a1b399f26304ce5342896d23680acbf878/startup/20-area-detectors.py#L160-L192

Also, that Eiger has a special file-writing functionality that is unlike the normal Area Detector file plugins. We wrote a one-off ophyd class to shoe-horn it into looking like ophyd's concept of a file plugin. 

https://github.com/NSLS-II-CHX/profile_collection/blob/7ef674a1b399f26304ce5342896d23680acbf878/startup/20-area-detectors.py#L117-L157

Would that be useful upstream? Do _all_ Eigers have this custom file-writing functionality? I have never been clear on that; we just dashed this CHX implementation out under pressure in very early days and haven't circled back to carefully consider it.

Reply:
I started with the CHX profile collection implementation as a base (the variable names are the same for the Signals shared with that implementation), so it should be a superset of that class. I just added signals for all the remaining PVs defined by the Eiger driver.

The file saving in Eiger that doesn't use the default file plugins is common to all Eiger detectors. Basically, since the detector is too fast to follow the normal plugin chain, the vendor API writes to a super-fast cache first, and then this built in file plugin in the eiger driver asynchronously copies the saved hdf5 files from the cache to the target location specified by the filepath PV.

This actually isn't limited to just the Eiger driver - there are several detectors that use File PVs to control built in vendor API file saving operations, skipping the typical areaDetector plugin chain. The giveaway is basically if the `NDFile.template` file is included in the driver database definition:

https://github.com/areaDetector/ADEiger/blob/01bb57cc2d1e3d78fbf56f88c38406be3e61574e/eigerApp/Db/eiger.template#L6

Basically to account for this, in a previous PR I had split the definition of the PVs defined in `NDFile.template` out from the `FilePlugin` class into the new `FileBase` class, and these signals should cover you for the one-off file plugin wrapper signals - they will now just be under the `EigerDetectorCam` umbrella, so that one-off class should theoretically be no longer needed. Instead, the stage/unstage code should probably be moved into a subclass of `EigerDetector`.

I think that overall what would be the most beneficial here in upstream is code to avoid the boilerplating required to create a usable ophyd object for an areaDetector, I believe described in an issue here: https://github.com/bluesky/ophyd/issues/889 and https://github.com/bluesky/ophyd/issues/888

Generally all areaDetectors use the same plugin format, so maybe helper functions that automatically add all the plugins loaded by default by [`commonPlugins.cmd`](https://github.com/areaDetector/ADCore/blob/master/iocBoot/EXAMPLE_commonPlugins.cmd) would be a nice start?

Reply:
Great. Thanks for explaining; I hadn't yet learned some of that context.

I suspected that your `FileBase` PR was laying track for this exact scenario but didn't want to assume.

I agree that would be a nice start. I am hopeful that DAMA can prioritize #888 and #889 soon, but if you have motivation and time to take it on yourself before then, please don't hesitate to claim one or both.